### ISSUE TITLE
CI: Flaky ResourceWarning('unclosed file <_io.BufferedRandom ... 

### ISSUE THREAD

**Issue by mroeschke** [pandas-dev/pandas#44844]
Example:
https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=70478&view=logs&jobId=cb04ab33-336b-54ee-68fe-dcd3abde835b&j=cb04ab33-336b-54ee-68fe-dcd3abde835b&t=0558d00d-ff4a-5538-0de9-a0e7f449700f

Previous debugging attempt: https://github.com/pandas-dev/pandas/pull/44634

Since it's a `io.BufferedRandom`, it should be a file handle opened in `+` mode: https://github.com/pandas-dev/pandas/pull/44634#issuecomment-981260080

This only appears when `tm.assert_produces_warning` is called since it's checking for extra warnings. It's likely coming from an optional dependency at this point, but unconfirmed.

**Comment by twoertwein** [comment]
maybe related https://github.com/pandas-dev/pandas/pull/44761#issuecomment-991355655

**Comment by mroeschke** [comment]
> maybe related [#44761 (comment)](https://github.com/pandas-dev/pandas/pull/44761#issuecomment-991355655)

Thanks, but I believe we've been seeing this error in the CI prior to that PR.

**Comment by twoertwein** [comment]
might be helpful for the next attempt at ResourceWarnings: my previous attempt at it #39047

**Comment by jbrockmendel** [comment]
Logging these warnings when they pop up

2021-12-15, #44908, py39GHA, `test_dialect_conflict_delimiter`, `<_io.BufferedRandom name=13>`, `(pandas/tests/io/parser/conftest.py', 22)`

**Comment by mroeschke** [comment]
Found a suspect from https://github.com/pandas-dev/pandas/runs/4540629821?check_suite_focus=true

Think it's related to https://github.com/matplotlib/matplotlib/issues/21646#issuecomment-982172602

TLDR from the log: `mpl-data/fonts/ttf/DejaVuSans.ttf` was opened an not appeared closed at OS number 14

```
2021-12-15T06:06:54.3616928Z [gw1] PASSED pandas/tests/libs/test_hashtable.py::TestHashTable::test_lookup[False-Int64HashTable-int640] python      5650 runner   13r      REG                8,1   756072 1827313 /usr/share/miniconda/envs/pandas-dev/lib/python3.8/site-packages/matplotlib/mpl-data/fonts/ttf/DejaVuSans.ttf
2021-12-15T06:06:54.3620174Z python      5650 runner   14r      REG                8,1   756072 1827313 /usr/share/miniconda/envs/pandas-dev/lib/python3.8/site-packages/matplotlib/mpl-data/fonts/ttf/DejaVuSans.ttf
```


```
2021-12-15T06:57:29.3149376Z         if extra_warnings:
2021-12-15T06:57:29.3149999Z >           raise AssertionError(f"Caused unexpected warning(s): {repr(extra_warnings)}")
2021-12-15T06:57:29.3151905Z E           AssertionError: Caused unexpected warning(s): [('ResourceWarning', ResourceWarning('unclosed file <_io.BufferedRandom name=14>'), '/usr/share/miniconda/envs/pandas-dev/lib/python3.8/site-packages/openpyxl/descriptors/serialisable.py', 52)]
```

**Comment by mroeschke** [comment]
If matplotlib is related, I am not sure there's any way to avoid this besides blocking the reading of some font files.

After investigating, I think at import time this object is reading certain files and might be leaving them open: https://github.com/matplotlib/matplotlib/blob/f6e0ee49c598f59c6e6cf4eefe473e4dc634a58a/src/ft2font.cpp#L322

**Comment by mroeschke** [comment]
So far the response from matplotlib is "this is the intended behavior". Opened up https://github.com/matplotlib/matplotlib/issues/22017 for a hopeful resolution to this issue. 

I don't have many ideas how to reliably handle this case in the meantime

1. Avoid all font files with `'otf', 'ttc', 'ttf'` extensions from being read. Matplotlib tests will still run.
2. Have matplotlib tests run last such that the matpliotlib import (and therefore font loading) is delayed until after all the `tm.assert_produces_warning` checks are done.

Anyone else have suggestions?

**Comment by jbrockmendel** [comment]
> Anyone else have suggestions?

Brainstorming:

3) Ask @pganssle about where to file an issue (BPO?) or open discussion about making these easier to track down
4) Open an issue with psutil about getting proc.open_files to identify these objects
5) IIUC the warning is issued when `__del__` is called before `close`.  Could tinker with `gc.collect` to
     a) raise instead of warn?
     b) disable collection and maybe then proc.open_files will find them if there's a race condition?
6) Run a few hundred builds without matplotlib installed at all and see if we can confirm the hypothesis these files are the source of the problem
     a) An automated way to "run a few hundred builds and collect the results" would be helpful for many of the CI troubles.
7) Throw in the towel and have tm.assert_produces_warning ignore ResourceWarnings
    a) Could have them ignore _only_ if matplotlib is found in sys.modules?

**Comment by jbrockmendel** [comment]
Another idea:

Add a check in `assert_produces_warning` along the lines of `if "--ignore-resource-warning in sys.argv[1:]": [...]` and then use that in some or all of the CI runs.

**Comment by jbrockmendel** [comment]
Have t seen this in a long time, did we just ignore it or did it resolve itself?

**Comment by mroeschke** [comment]
They're probably either resolved or suppressed but might as well close
